x-pack/plugins/task_manager/server/queries/task_claiming.ts

```diff
@@ -0,0 +1,488 @@
+  claimOwnershipUntil: Date;
+  claimTasksById?: string[];
+  size: number;
+  taskTypes: Set<string>;
+}
+export type IncrementalOwnershipClaimingOpts = OwnershipClaimingOpts & {
+  precedingQueryResult: UpdateByQueryResult;
+};
+export type IncrementalOwnershipClaimingReduction = (
+  opts: IncrementalOwnershipClaimingOpts
+) => Promise<UpdateByQueryResult>;
+
+export interface FetchResult {
+  docs: ConcreteTaskInstance[];
+}
+
+export interface ClaimOwnershipResult {
+  stats: {
+    tasksUpdated: number;
+    tasksConflicted: number;
+    tasksClaimed: number;
+    tasksRejected: number;
+  };
+  docs: ConcreteTaskInstance[];
+  timing?: TaskTiming;
+}
+
+enum BatchConcurrency {
+  Unlimited,
+  Limited,
+}
+
+type TaskClaimingBatches = Array<UnlimitedBatch | LimitedBatch>;
+interface TaskClaimingBatch<Concurrency extends BatchConcurrency, TaskType> {
+  concurrency: Concurrency;
+  tasksTypes: TaskType;
+}
+type UnlimitedBatch = TaskClaimingBatch<BatchConcurrency.Unlimited, Set<string>>;
+type LimitedBatch = TaskClaimingBatch<BatchConcurrency.Limited, string>;
+
+export class TaskClaiming {
+  public readonly errors$ = new Subject<Error>();
+  public readonly maxAttempts: number;
+
+  private definitions: TaskTypeDictionary;
+  private events$: Subject<TaskClaim>;
+  private taskStore: TaskStore;
+  private getCapacity: (taskType?: string) => number;
+  private logger: Logger;
+  private readonly taskClaimingBatchesByType: TaskClaimingBatches;
+  private readonly taskMaxAttempts: Record<string, number>;
+
+  /**
+   * Constructs a new TaskStore.
+   * @param {TaskClaimingOpts} opts
+   * @prop {number} maxAttempts - The maximum number of attempts before a task will be abandoned
+   * @prop {TaskDefinition} definition - The definition of the task being run
+   */
+  constructor(opts: TaskClaimingOpts) {
+    this.definitions = opts.definitions;
+    this.maxAttempts = opts.maxAttempts;
+    this.taskStore = opts.taskStore;
+    this.getCapacity = opts.getCapacity;
+    this.logger = opts.logger;
+    this.taskClaimingBatchesByType = this.partitionIntoClaimingBatches(this.definitions);
+    this.taskMaxAttempts = Object.fromEntries(this.normalizeMaxAttempts(this.definitions));
+
+    this.events$ = new Subject<TaskClaim>();
+  }
+
+  private partitionIntoClaimingBatches(definitions: TaskTypeDictionary): TaskClaimingBatches {
+    const {
+      limitedConcurrency,
+      unlimitedConcurrency,
+      skippedTypes,
+    } = groupBy(definitions.getAllDefinitions(), (definition) =>
+      definition.maxConcurrency
+        ? 'limitedConcurrency'
+        : definition.maxConcurrency === 0
+        ? 'skippedTypes'
+        : 'unlimitedConcurrency'
+    );
+
+    if (skippedTypes?.length) {
+      this.logger.info(
+        `Task Manager will never claim tasks of the following types as their "maxConcurrency" is set to 0: ${skippedTypes
+          .map(({ type }) => type)
+          .join(', ')}`
+      );
+    }
+    return [
```

Comment:
a thought: if we shuffled these randomly - including the unlimited concurrency ones - I'd guess when we have a stampede herd of Kibanas starting, there will less of a chance of optimistic concurrency version conflicts, since there's a good chance each Kibana might be searching over different concurrency groups.

Will be a fun experiment after we ship this :-)

Reply:
haha we can try but I suspect its impact would be minimal... 
The shifting mechanism kicks in if queries clash and the latest fix there has made it quite stable (far fewer cascades than we saw in 7.11). Adding shuffle here could actually add _more_ chaos to this that doesn't feel like it'll help.. but we can give it a shot if you'd like 🤷 

My personal feeling is that we've pushed this approach as far as it will go and we should explore more robust approaches (coordinating nodes etc.).